File: binary_tree/Java/VerticalSum.java
class ListNode
{
    int data;
    ListNode prev, next;

    ListNode(int data, ListNode prev, ListNode next)
    {
        this.data = data;
        this.prev = prev;
        this.next = next;
    }
}

public class VerticalSum {
    public static void VerticalSum(Node root, ListNode curr) {
        if (root == null) {
            return;
        }
        curr.data += root.data;

        if (root.left != null && curr.prev == null) {
            curr.prev = new ListNode(0, null, curr);
        }

        if (root.right != null && curr.next == null) {
            curr.next = new ListNode(0, curr, null);
        }

        VerticalSum(root.left, curr.prev);
        VerticalSum(root.right, curr.next);
    }
    public static void print(ListNode mid)
    {
        // find the head node
        while (mid != null && mid.prev != null) {
            mid = mid.prev;
        }

        ListNode head = mid;
        while (head != null)
        {
            System.out.print(head.data + " ");
            head = head.next;
        }
    }
    public static void printVerticalSum(Node root)
    {
        // base case
        if (root == null) {
            return;
        }

        ListNode curr = new ListNode(0, null, null);

        VerticalSum(root, curr);

        // print the linked list
        print(curr);
    }

    public static void main(String[] args) {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.right.left = new Node(5);
        root.right.right = new Node(6);
        root.right.left.left = new Node(7);
        root.right.left.right = new Node(8);

        printVerticalSum(root);
    }
}

File: binary_tree/python/verticalsum.py
class LNode:
    def __init__(self, val, prev = None, nnext = None):
        self.val = val
        self.prev = prev
        self.next = nnext

class Node:
    def __init__(self, val):
        self.val = val
        self.left = self.right = None

def vertical_sum(root, node):
    if not root:    return root
    node.val += root.val
    if root.left and not node.prev:    node.prev = LNode(0, None, node)
    vertical_sum(root.left, node.prev)
    if root.right and not node.next:    node.next = LNode(0, node, None)
    vertical_sum(root.right, node.next)

if __name__=='__main__':
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.right.left = Node(5)
    root.right.right = Node(6)
    root.right.left.left = Node(7)
    root.right.left.right = Node(8)
    head = LNode(0)
    vertical_sum(root, head)
    while head.prev:
        head = head.prev
    while head:
        print(head.val, end = ' ')
        head = head.next